src/bio/bio_buffer.c

```diff
@@ -681,105 +743,71 @@ dma_rw(struct bio_desc *biod, bool prep)
-		pg_idx = rg->brr_off >> BIO_DMA_PAGE_SHIFT;
-		payload = rg->brr_chk->bdc_ptr +
-			(rg->brr_pg_idx << BIO_DMA_PAGE_SHIFT);
-
-		if (!rmw_read) {
-			pg_cnt = (rg->brr_end + BIO_DMA_PAGE_SZ - 1) >>
-					BIO_DMA_PAGE_SHIFT;
-			D_ASSERT(pg_cnt > pg_idx);
-			pg_cnt -= pg_idx;
-
-			biod->bd_inflights++;
-
-			D_DEBUG(DB_IO, "%s blob:%p payload:%p, "
-				"pg_idx:"DF_U64", pg_cnt:"DF_U64"\n",
-				biod->bd_update ? "Write" : "Read",
-				blob, payload, pg_idx, pg_cnt);
-
-			if (biod->bd_update)
-				spdk_blob_io_write(blob, channel, payload,
-					page2io_unit(biod->bd_ctxt, pg_idx),
-					page2io_unit(biod->bd_ctxt, pg_cnt),
-					rw_completion, biod);
-			else
-				spdk_blob_io_read(blob, channel, payload,
-					page2io_unit(biod->bd_ctxt, pg_idx),
-					page2io_unit(biod->bd_ctxt, pg_cnt),
-					rw_completion, biod);
-			continue;
-		}
-
-		/*
-		 * Since DAOS doesn't support partial overwrite yet, we don't
-		 * do RMW for partial update, only zeroing the page instead.
-		 */
-		pg_off = rg->brr_off & ((uint64_t)BIO_DMA_PAGE_SZ - 1);
-
-		if (pg_off != 0 && payload != pg_rmw) {
-			D_DEBUG(DB_IO, "Front partial blob:%p payload:%p, "
-				"pg_idx:"DF_U64" pg_off:%d\n",
-				blob, payload, pg_idx, pg_off);
-
-			memset(payload, 0, BIO_DMA_PAGE_SZ);
-			pg_rmw = payload;
-		}
+		D_ASSERT(rg->brr_end > rg->brr_off);
 
-		pg_end = rg->brr_end >> BIO_DMA_PAGE_SHIFT;
-		D_ASSERT(pg_end >= pg_idx);
-		payload += (pg_end - pg_idx) << BIO_DMA_PAGE_SHIFT;
-		pg_off = rg->brr_end & ((uint64_t)BIO_DMA_PAGE_SZ - 1);
-
-		if (pg_off != 0 && payload != pg_rmw) {
-			D_DEBUG(DB_IO, "Rear partial blob:%p payload:%p, "
-				"pg_idx:"DF_U64" pg_off:%d\n",
-				blob, payload, pg_idx, pg_off);
-
-			memset(payload, 0, BIO_DMA_PAGE_SZ);
-			pg_rmw = payload;
-		}
-	}
-
-	if (xs_ctxt->bxc_tgt_id == -1) {
-		D_DEBUG(DB_IO, "Self poll completion, blob:%p\n", blob);
-		xs_poll_completion(xs_ctxt, &biod->bd_inflights);
-	} else {
-		biod->bd_dma_issued = 1;
+		if (rg->brr_media == DAOS_MEDIA_SCM)
+			scm_rw(biod, rg);
+		else
+			nvme_rw(biod, rg);
+ 	}
+ 
+ 	if (xs_ctxt->bxc_tgt_id == -1) {
+		D_DEBUG(DB_IO, "Self poll completion\n");
+ 		xs_poll_completion(xs_ctxt, &biod->bd_inflights);
+ 	} else {
+ 		biod->bd_dma_issued = 1;
 		if (biod->bd_inflights != 0)
 			ABT_eventual_wait(biod->bd_dma_done, NULL);
-	}
-
-	biod->bd_ctxt->bic_inflight_dmas--;
-	D_DEBUG(DB_IO, "DMA done, blob:%p, update:%d, rmw:%d\n",
-		blob, biod->bd_update, rmw_read);
-}
-
-void
-bio_memcpy(struct bio_desc *biod, uint16_t media, void *media_addr,
-	   void *addr, ssize_t n)
-{
-	struct umem_instance *umem = biod->bd_ctxt->bic_umem;
-
-	if (biod->bd_update && media == DAOS_MEDIA_SCM) {
-		/* NB: pmemobj_tx_commit will drain HW buffer */
-		pmemobj_memcpy(umem->umm_pool, media_addr, addr, n,
-			       PMEMOBJ_F_MEM_NODRAIN);
-	} else {
-		if (biod->bd_update)
-			memcpy(media_addr, addr, n);
-		else
-			memcpy(addr, media_addr, n);
-	}
+ 	}
+ 
+ 	biod->bd_ctxt->bic_inflight_dmas--;
```

Comment:
(style)  code indent should use tabs where possible
(style)  please, no space before tabs